src/main/java/za/co/neslotech/spring/trainspringbootinsight/controller/CrudController.java

```diff
@@ -0,0 +1,102 @@
+package za.co.neslotech.spring.trainspringbootinsight.controller;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.EntityNotFoundException;
+import jakarta.servlet.http.HttpServletRequest;
+import org.springframework.beans.BeanUtils;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A generic REST API Controller with CRUD operations.
+ * The child class should use the following annotations
+ *      1. @RestController
+ * @param <T> The entity class.
+ * @param <I> The primary key field type.
+ */
+public abstract class CrudController<T, I> {
+
+    protected abstract String[] getUpdateColumnExclusions();
+    protected abstract String getEntityId(T entity);
+    protected abstract T saveAndFlush(T entity);
+    protected abstract List<T> findAll();
+    protected abstract Optional<T> findById(I id);
+    protected abstract void delete(T entity);
```

Comment:
We can simplify this quite drastically, I would remove these and do the following:

```suggestion
    protected abstract CrudService getService();
```

This will mean you need to create an abstract CrudService (could even be an interface) that will then contain the following functions:

1. findAll(): List<T>
2. findById(final I id): T
3. update(final I id, T entity): T
4. delete(final I id): void